lib/Dialect/Verif/Transforms/LowerContracts.cpp

```diff
@@ -0,0 +1,109 @@
+//===- LowerContracts.cpp - Formal Preparations --*- C++ -*----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Lower contracts into verif.formal tests.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Dialect/Verif/VerifPasses.h"
+
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace circt;
+
+namespace circt {
+namespace verif {
+#define GEN_PASS_DEF_LOWERCONTRACTSPASS
+#include "circt/Dialect/Verif/Passes.h.inc"
+} // namespace verif
+} // namespace circt
+
+using namespace mlir;
+using namespace verif;
+using namespace hw;
+
+namespace {
+struct LowerContractsPass
+    : verif::impl::LowerContractsPassBase<LowerContractsPass> {
+  void runOnOperation() override;
+};
+
+template <typename FROM, typename TO>
+void replaceContractOp(PatternRewriter &rewriter, Block *body) {
+  for (auto op : llvm::make_early_inc_range(body->getOps<FROM>())) {
```

Comment:
Might want a walk here or error out if there's an op with a region that could contain an op to convert.